File: 113.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    vector<vector<int>> res;
    void dfs(TreeNode* root,vector<int> &tmp,int sum){
        if(!root) return ;
        tmp.push_back(root->val);
        if(root->val==sum&&!root->left&&!root->right){
            res.push_back(tmp);
        }
        dfs(root->left,tmp,sum-root->val);
        dfs(root->right,tmp,sum-root->val);
        tmp.pop_back();
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<int> tmp;
        dfs(root,tmp,targetSum);
        return res;
    }
};

File: 113.java
import java.util.*;
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
};

class Solution113 {
    List<List<Integer>> res = new LinkedList<List<Integer>>();
    List<Integer> path = new LinkedList<Integer>();
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        dfs(root, targetSum);
        return res;
    }
    public void dfs(TreeNode root, int targetSum){
        if(root==null)return;
        path.add(root.val);
        if(root.val == targetSum && root.left==null && root.right==null){
            res.add(new LinkedList<Integer>(path));
        }
        dfs(root.left, targetSum-root.val);
        dfs(root.right, targetSum-root.val);
        path.remove(path.size()-1);
    }
}
